Framework/DataHandling/src/SaveDiffCal.cpp

```diff
@@ -169,27 +213,56 @@ void SaveDiffCal::generateDetidToIndex() {
   }
 }
 
+void SaveDiffCal::generateDetidToIndex(const DataObjects::SpecialWorkspace2D_const_sptr &ws) {
+  if (!bool(ws))
+    throw std::runtime_error(
+        "Encountered null pointer in SaveDiffCal::generateDetidToIndex which should be impossible");
+
+  m_detidToIndex.clear();
+
+  std::size_t index = 0;
+  for (size_t i = 0; i < m_numValues; ++i) {
+    const auto &detids = ws->getSpectrum(i).getDetectorIDs();
+    for (const auto &detid : detids) {
+      m_detidToIndex[static_cast<detid_t>(detid)] = index;
+      index++;
+    }
+  }
+}
+
 bool SaveDiffCal::tableHasColumn(const std::string &ColumnName) const {
-  const std::vector<std::string> names = m_calibrationWS->getColumnNames();
-  return std::any_of(names.cbegin(), names.cend(), [&ColumnName](const auto &name) { return name == ColumnName; });
+  if (m_calibrationWS) {
+    const std::vector<std::string> names = m_calibrationWS->getColumnNames();
+    return std::any_of(names.cbegin(), names.cend(), [&ColumnName](const auto &name) { return name == ColumnName; });
+  } else {
+    return false;
+  }
 }
 
 //----------------------------------------------------------------------------------------------
 /** Execute the algorithm.
  */
 void SaveDiffCal::exec() {
   m_calibrationWS = getProperty("CalibrationWorkspace");
-  this->generateDetidToIndex();
-
   GroupingWorkspace_sptr groupingWS = getProperty("GroupingWorkspace");
-  if (bool(groupingWS) && groupingWS->isDetectorIDMappingEmpty())
-    groupingWS->buildDetectorIDMapping();
-
   MaskWorkspace_const_sptr maskWS = getProperty("MaskWorkspace");
 
+  if (m_calibrationWS) {
+    m_numValues = m_calibrationWS->rowCount();
+    this->generateDetidToIndex();
+  } else if (groupingWS) {
+    m_numValues = groupingWS->getNumberHistograms();
+    this->generateDetidToIndex(groupingWS);
+  } else if (maskWS) {
+    m_numValues = maskWS->getNumberHistograms();
+    this->generateDetidToIndex(maskWS);
+  }
```

Comment:
This new block of code deals with initializing `m_numValues`. But so does L264-L268. Can these blocks of code be consolidated?

Reply:
These get a starting value. The other block finds out what the minimum number of values are.

Reply:
I see, but I still think the code would be easier to read if L261-262 (calling `buildDetectorIDMapping`) were executed _after_ all the issues with `m_numValues` are sorted out.
Also the comment 
```
// initialize `m_numValues` as the minimum of (CalibrationWorkspace_row_count,
  // GroupingWorkspace_histogram_count, MaskWorkspace_histogram_count)
```
was written when CalibrationWorkspace was always there, so I think it needs to be revised.

Reply:
```suggestion
  m_numValues = 0;
  if (m_calibrationWS) {
    m_numValues = m_calibrationWS->rowCount();
    this->generateDetidToIndex();
  }
  
  if (groupingWS) {
    if(m_numValues > 0)
      m_numValues = std::min(m_numValues, groupingWS->getNumberHistograms()) ;
   this->generateDetidToIndex(groupingWS);
  }
  
  if (maskWS) {
    if(m_numValues > 0)
      m_numValues = std::min(m_numValues, maskWS->getNumberHistograms()) ;
    this->generateDetidToIndex(maskWS);
  }
```

Reply:
Then there is no need for the second (old) block of code.

Reply:
the `generateDetidToIndex` calls with `GroupingWorkspace` and `MaskWorkspace` should only happen if the calibration workspace doesn't exist. I'm staring at it a little more to see if I can fix that issue because your version does have some other good points to it.